<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Week 3 - Working with data in R</title>

<script src="site_libs/header-attrs-2.20/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img style="width: 120px; float: left; margin-top: 0px; margin-right: 20px;" src="images/CRUK_CI_logo.png"/>R for Biologists&nbsp;&nbsp;&nbsp;</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Contents
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="week1.html">Week 1 - An introduction to RStudio, R objects, data types, functions and vectors</a>
    </li>
    <li>
      <a href="week2.html">Week 2 - An introduction to other data structures</a>
    </li>
    <li>
      <a href="week3.html">Week 3 - Working with data in R</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/bioinformatics-core-shared-training/Bitesize-R">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Week 3 - Working with data in R</h1>

</div>


<blockquote>
<h4 id="learning-objectives">Learning objectives</h4>
<ul>
<li>Create an R Project</li>
<li>Learn how to install and use packages that extend R’s basic
functionality</li>
<li>Read data in tabular format into R</li>
<li>Calculate summary statistics on your tabular data</li>
<li>Introduce the <strong><code>tibble</code></strong>, arguably the
most important data structure you will use in R</li>
<li>Learn how R deals with missing values</li>
<li>Learn the two key <strong>dplyr</strong> functions for manipulating
your data
<ul>
<li><strong><code>filter()</code></strong> for selecting observations
based on their values, i.e. selecting rows in your table</li>
<li><strong><code>select()</code></strong> for selecting a subset of
variables, i.e. selecting columns in your table</li>
<li><strong><code>arrange()</code></strong> for sorting the observations
in your table</li>
</ul></li>
</ul>
</blockquote>
<div id="create-an-r-project" class="section level2">
<h2>Create an R Project</h2>
<p>Before starting to write code in RStudio, we need to create an R
Project. The idea behind an R project is to have a workspace where you
can keep all the files and settings associated with the project
together. In that way, next time you open the R Project it would be
easier to resume work.</p>
<p>To create an “R Project”:</p>
<ol style="list-style-type: decimal">
<li>Start RStudio.</li>
<li>Under the <code>File</code> menu, click on <code>New Project</code>.
Choose <code>New Directory</code>, then <code>New Project</code>.</li>
<li>Enter a name for this new folder (or “directory”), and choose a
convenient location for it. This will be your <strong>working
directory</strong> for the rest of the day (<em>e.g.,</em>
<code>~/my_first_project</code>).</li>
<li>Click on <code>Create Project</code>.</li>
<li>(Optional) Set Preferences to ‘Never’ save workspace in
RStudio.</li>
</ol>
<p>RStudio’s default preferences generally work well, but saving a
workspace to <code>.RData</code> can be cumbersome, especially if you
are working with larger datasets as this would save all the data that is
loaded into R into the <code>.RData</code> file.<br />
To turn that off, go to <code>Tools</code> –&gt;
<code>Global Options</code> and select the ‘Never’ option for
<code>Save workspace to .RData' on exit.</code></p>
</div>
<div id="getting-started-with-data-in-r" class="section level1">
<h1>Getting started with data in R</h1>
<p>In this course, we’ll be mostly focusing on a set of R packages
specifically designed for the most useful and common operations for
interacting with and visualizing data, and particularly data in a
tabular form. This collection of packages is known as the
<strong>tidyverse</strong>.</p>
<p>Learning the <strong>tidyverse</strong> is more than just learning
about some helpful functions to get certain tasks done. The tidyverse
packages form a coherant system for data manipulation, exploration and
visualization that share a common design philosophy. There is a certain
elegance to code crafted using the tidyverse approach that is in stark
contrast to the strange and often cryptic equivalent using traditional
‘base R’.</p>
<p>We will also start to look at the most important data structure
you’ll use with your data, assuming it is in tabular form, the
<strong>data frame</strong>, and its superior tidyverse derivative, the
<strong>tibble</strong>.</p>
</div>
<div id="packages" class="section level1">
<h1>Packages</h1>
<p>Before we move on to the tidyverse version of a data frame, the
tibble, we have to load the tidyverse package that provides the tibble
functionality. It doesn’t come with the set of packages loaded when you
first start R.</p>
<p>There are thousands of R packages that extend R’s core functionality
to provide specialized functions for doing all kinds of things. There
are R packages that provide various kinds of visualizations and plotting
functions, others that let us manipulate data in interesting and useful
ways, and still others that give us access to cutting-edge machine
learning algorithms.</p>
<p>You need to load a package using the
<strong><code>library()</code></strong> function in order to be able to
use the functions and data sets that the package provides. The tibble
data structure is part of the <strong><code>tibble</code></strong>
package so to make this available you would run the following
command:</p>
<pre class="r"><code>library(tibble)</code></pre>
<p>The <code>tibble</code> package also contains various useful
functions for converting other data structures, e.g. data frames, to
tibbles and for working with tibble objects.</p>
<p>The <code>tibble</code> package is one of a collection of packages
that make up the tidyverse. It is usually more convenient to load the
core tidyverse packages in one go as follows.</p>
<pre class="r"><code>library(tidyverse)</code></pre>
<pre><code>## ── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
## ✔ dplyr     1.1.0     ✔ readr     2.1.4
## ✔ forcats   1.0.0     ✔ stringr   1.5.0
## ✔ ggplot2   3.4.1     ✔ tibble    3.2.0
## ✔ lubridate 1.9.2     ✔ tidyr     1.3.0
## ✔ purrr     1.0.1     
## ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
## ✖ dplyr::filter() masks stats::filter()
## ✖ dplyr::lag()    masks stats::lag()
## ℹ Use the ]8;;http://conflicted.r-lib.org/conflicted package]8;; to force all conflicts to become errors</code></pre>
<p>The output from this command tells us which packages were loaded, one
of which is the <code>tibble</code> package. Another of these is
<code>ggplot2</code> that provides excellent plotting functions for
visualizing our data and will be the main focus next week.</p>
<p>We also got a warning about some conflicts. This tells us that there
are functions loaded as part of the <code>dplyr</code> package that have
the same name as functions that are in the <code>stats</code> package.
The <code>stats</code> package was already loaded, from when we first
started our R session, and the functions referred to in the warning are
now masked; to use these instead of their <code>dplyr</code> namesakes
you need to specify them fully by adding the package name and
<code>::</code> as shown in the warning message.</p>
<div id="installing-packages" class="section level2">
<h2>Installing packages</h2>
<p>If you got the following error message when trying to load the
tidyverse, then you haven’t yet installed it.</p>
<pre><code>&gt; library(tidyverse)
Error in library(tidyverse) : there is no package called ‘tidyverse’</code></pre>
<p>Installing packages should be relatively straightforward using the
<code>install.packages()</code> function.</p>
<pre class="r"><code>install.packages(&quot;tidyverse&quot;)</code></pre>
<p>If you’re using R at a regular console within a terminal window,
i.e. not RStudio, you may be asked to select one of several mirrors of
the CRAN R package repository hosted at different locations around the
world (in the UK these are in Bristol and London).</p>
<p>For some packages you may be asked to choose between a binary and a
source package, particularly if the latest version of the package is
only available in source form and that package contains code written in
another language such as C. This could be a problem if you don’t have a
C compiler installed on your computer.</p>
</div>
</div>
<div id="tibbles" class="section level1">
<h1>Tibbles</h1>
<p>The tidyverse brings us a special type of data frame called the
<strong><code>tibble</code></strong>.</p>
<p>Data frames have some rather quirky and annoying behaviours and the
<code>tibble</code> addresses some of these issues to make life a little
easier. We’ll first look at how we can create a tibble and then will
have a quick look at some of the advantages of the tibble over the data
frame.</p>
<div id="creating-a-tibble" class="section level2">
<h2>Creating a tibble</h2>
<p>Tibbles can be created using the <code>tibble()</code> function in a
similar manner to how we created a data frame earlier.</p>
<pre class="r"><code>beatles &lt;- tibble(
  name = c(&quot;John&quot;, &quot;Paul&quot;, &quot;Ringo&quot;, &quot;George&quot;),
  birth_year = c(1940, 1942, 1940, 1943),
  instrument = c(&quot;guitar&quot;, &quot;bass&quot;, &quot;drums&quot;, &quot;guitar&quot;)
)
beatles</code></pre>
<pre><code>## # A tibble: 4 × 3
##   name   birth_year instrument
##   &lt;chr&gt;       &lt;dbl&gt; &lt;chr&gt;     
## 1 John         1940 guitar    
## 2 Paul         1942 bass      
## 3 Ringo        1940 drums     
## 4 George       1943 guitar</code></pre>
<p>A data frame can be converted into a tibble using the
<code>as_tibble()</code> function.</p>
<pre class="r"><code>iris_tibble &lt;- as_tibble(iris)</code></pre>
<p>Our iris tibble is still a data frame but it also has acquired some
additional types or classes (<code>tbl_df</code> and
<code>tbl</code>).</p>
<pre class="r"><code>class(iris)</code></pre>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
<pre class="r"><code>class(iris_tibble)</code></pre>
<pre><code>## [1] &quot;tbl_df&quot;     &quot;tbl&quot;        &quot;data.frame&quot;</code></pre>
</div>
<div id="tibbles-vs-data-frames" class="section level2">
<h2>Tibbles vs data frames</h2>
<p>So let’s now have a look at what capabilities and behaviours these
additional types confer and what the main differences are between
tibbles and data frames.</p>
<p>The first immediate difference from a <code>data.frame</code> is that
a <code>tibble</code> displays the data type of each column under its
name and it only prints as many columns as fit on one screen. Another
difference is that printing a <code>tibble</code> will not print the
whole dataset, but just the first 10 rows and only the columns that fit
the screen (same as <code>head</code> but with 10 rows instead of 6). If
you would like to print more than the first 10 rows use the
<code>print</code> function.</p>
<p>Since printing <code>tibble</code> already gives you information
about the data structure, the data types of each column and the size of
the dataset, the <code>str</code> function is not as much useful as it
was when using <code>data.frame</code>.</p>
<pre class="r"><code>## inspect the data
str(iris_tibble)</code></pre>
<pre><code>## tibble [150 × 5] (S3: tbl_df/tbl/data.frame)
##  $ Sepal.Length: num [1:150] 5.1 4.9 4.7 4.6 5 5.4 4.6 5 4.4 4.9 ...
##  $ Sepal.Width : num [1:150] 3.5 3 3.2 3.1 3.6 3.9 3.4 3.4 2.9 3.1 ...
##  $ Petal.Length: num [1:150] 1.4 1.4 1.3 1.5 1.4 1.7 1.4 1.5 1.4 1.5 ...
##  $ Petal.Width : num [1:150] 0.2 0.2 0.2 0.2 0.2 0.4 0.3 0.2 0.2 0.1 ...
##  $ Species     : Factor w/ 3 levels &quot;setosa&quot;,&quot;versicolor&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...</code></pre>
<p>Notice that rather than specifing <code>tibble</code> as the data
structure of surveys, the first line of <code>str</code>’s output now
specifies <code>‘spec_tbl_df’, ‘tbl_df’, ‘tbl’ and 'data.frame'</code>
which can be a bit confusing. These are the classes <code>tibble</code>
inherts from which in simple terms means that <code>tibble</code> is a
<code>data.frame</code> with a few modifications. Therefore, most of the
functions that were used with <code>data.frame</code> can also be used
with <code>tibble</code>.</p>
</div>
</div>
<div id="reading-data-into-r" class="section level1">
<h1>Reading data into R</h1>
<p>Another of the tidyverse packages loaded when we ran
<code>library(tidyverse)</code> is the <code>readr</code> package, which
provides functions for reading tabular data into a tibble and writing
tibbles to tab-delmited and comma-separated value (CSV) files.</p>
<p>For example, the clinical data for the METABRIC breast cancer dataset
generated at CRUK CI can be downloaded as a tab-delimited file from <a
href="https://www.cbioportal.org/study/clinicalData?id=brca_metabric">cBioPortal</a>.
Targeted sequencing was performed for 2509 primary breast tumours, along
with 548 matched normals, using a panel of 173 of the most frequently
mutated breast cancer genes as part of the METABRIC study (<a
href="https://www.ncbi.nlm.nih.gov/pubmed/27161491">Pereira <em>et
al.</em>, Nature Communications 7:11479, 2016</a>).</p>
<p>Let us first download the file and have a look at the data.</p>
<p>We are going to use the R function <code>download.file()</code> to
download the TSV/CSV files that contains the data.</p>
<pre class="r"><code>download.file(url=&quot;https://bioinformatics-core-shared-training.github.io/Bitesize-R/data/brca_metabric_clinical_data.tsv&quot;,destfile = &quot;data/brca_metabric_clinical_data.tsv&quot;)
download.file(url=&quot;https://bioinformatics-core-shared-training.github.io/Bitesize-R/data/metabric_clinical_and_expression_data.csv&quot;,destfile = &quot;data/metabric_clinical_and_expression_data.csv&quot;)</code></pre>
<p>This file contains information in tabular form in which values are
separated by tab characters so we’ll use the
<strong><code>read_tsv()</code></strong> to read these data.</p>
<pre class="r"><code>metabric &lt;- read_tsv(&quot;data/brca_metabric_clinical_data.tsv&quot;)</code></pre>
<pre><code>## Rows: 2509 Columns: 20
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;\t&quot;
## chr (14): Patient ID, Sample ID, Cancer Type, Cancer Type Detailed, 3-Gene c...
## dbl  (6): Mutation Count, Age at Diagnosis, Cohort, Lymph nodes examined pos...
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<p><code>read_tsv()</code> prints a message telling us which types it
has assigned to each of the columns it has read, although in this case
the table has quite a lot of columns and it only outputs details about
the first few.</p>
<p>If we print the clinical data set we can see that
<code>read_tsv()</code> created a tibble.</p>
<pre class="r"><code>metabric</code></pre>
<pre><code>## # A tibble: 2,509 × 20
##    `Patient ID` Sample…¹ Cance…² Cance…³ Mutat…⁴ 3-Gen…⁵ Age a…⁶ Cellu…⁷ Chemo…⁸
##    &lt;chr&gt;        &lt;chr&gt;    &lt;chr&gt;   &lt;chr&gt;     &lt;dbl&gt; &lt;chr&gt;     &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;  
##  1 MB-0000      MB-0000  Breast… Breast…      NA ER-/HE…    75.6 &lt;NA&gt;    NO     
##  2 MB-0002      MB-0002  Breast… Breast…       2 ER+/HE…    43.2 High    NO     
##  3 MB-0005      MB-0005  Breast… Breast…       2 &lt;NA&gt;       48.9 High    YES    
##  4 MB-0006      MB-0006  Breast… Breast…       1 &lt;NA&gt;       47.7 Modera… YES    
##  5 MB-0008      MB-0008  Breast… Breast…       2 ER+/HE…    77.0 High    YES    
##  6 MB-0010      MB-0010  Breast… Breast…       4 ER+/HE…    78.8 Modera… NO     
##  7 MB-0014      MB-0014  Breast… Breast…       4 &lt;NA&gt;       56.4 Modera… YES    
##  8 MB-0020      MB-0020  Breast… Breast…      NA ER-/HE…    70   High    YES    
##  9 MB-0022      MB-0022  Breast… Breast…       1 &lt;NA&gt;       89.1 Modera… NO     
## 10 MB-0025      MB-0025  Breast… Breast…       5 &lt;NA&gt;       76.2 &lt;NA&gt;    &lt;NA&gt;   
## # … with 2,499 more rows, 11 more variables: Cohort &lt;dbl&gt;, `ER Status` &lt;chr&gt;,
## #   `ER status measured by IHC` &lt;chr&gt;, `HER2 Status` &lt;chr&gt;,
## #   `HER2 status measured by SNP6` &lt;chr&gt;, `Hormone Therapy` &lt;chr&gt;,
## #   `Inferred Menopausal State` &lt;chr&gt;, `Integrative Cluster` &lt;chr&gt;,
## #   `Lymph nodes examined positive` &lt;dbl&gt;, `Neoplasm Histologic Grade` &lt;dbl&gt;,
## #   `Nottingham prognostic index` &lt;dbl&gt;, and abbreviated variable names
## #   ¹​`Sample ID`, ²​`Cancer Type`, ³​`Cancer Type Detailed`, ⁴​`Mutation Count`, …</code></pre>
<div class="rmdblock">
<p>
<strong><code>read_csv()</code></strong> and
<strong><code>read_tsv()</code></strong>
</p>
<p>
<code>read_csv()</code> reads data from a comma-separated value (CSV)
file into a tibble. <code>read_tsv()</code> is the equivalent function
that works on tab-delmited files.
</p>
<p>
These functions expect the first line to contain column names and try to
make sensible guesses at the type of data in each column. You can change
this by specifying various arguments, e.g. to skip comment lines
beginning with a specific character (e.g. “#”) or to tell the function
what the column types are.
</p>
</div>
<p>To create clean column names we can use <code>clean_names()</code>
from <code>janitor</code> package.</p>
<pre class="r"><code>#install.packages(&quot;janitor&quot;)
library(janitor)</code></pre>
<pre><code>## 
## Attaching package: &#39;janitor&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     chisq.test, fisher.test</code></pre>
<pre class="r"><code>metabric &lt;- clean_names(metabric)</code></pre>
<p>We are now in a position to explore these data. For example, we can
use the <strong><code>table()</code></strong> function to count the
number of patients in this cohort that had chemotherapy and those that
did not.</p>
<pre class="r"><code>table(metabric$chemotherapy)</code></pre>
<pre><code>## 
##   NO  YES 
## 1568  412</code></pre>
<p>We can summarize the numbers of mutations found per patient.</p>
<pre class="r"><code>summary(metabric$mutation_count)</code></pre>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA&#39;s 
##   1.000   3.000   5.000   5.579   7.000  80.000     152</code></pre>
<div id="working-directory" class="section level2">
<h2>Working directory</h2>
<p>When reading in the METABRIC dataset we specified the file using a
relative path, <code>data/brca_metabric_clinical_data.tsv</code>. This
is relative to the current working directory and is in the
<code>data</code> subdirectory within our working directory. The
<strong><code>getwd()</code></strong> function will tell you what the
current working directory is.</p>
<pre class="r"><code>getwd()</code></pre>
<p>If you open RStudio by clicking on its icon in the task bar or its
entry in the start menu, your working directory will probably be your
home folder.</p>
<p>You can specify a full path for the file you want to read, which is
also known as an absolute path.</p>
<p>You can explicitly set your working directory using
<strong><code>setwd()</code></strong>.</p>
<p>If you’re running an R script from a terminal window (by typing
something like <code>Rscript my_script.R</code> at the command prompt)
your working directory will be the directory in which you run the script
(i.e. the one in which you issue the Rscript command). In this case you
will probably want to read files from locations that are relative to the
directory in which you run the script.</p>
<p>If you’re writing an R script in RStudio you can set the working
directory to be the same directory as that which contains your script
file using the menu: <strong>Session &gt; Set Working Directory &gt; To
Source File Location</strong>. You can also navigate to the directory of
your choosing in the Files tab pane and set this to be the working
directory by clicking on the <strong>More</strong> menu button and
selecting <strong>Set As Working Directory</strong>.</p>
</div>
</div>
<div id="missing-values" class="section level1">
<h1>Missing values</h1>
<p>The METABRIC clinical dataset has some missing values. If you look
back to the table we printed out earlier you can see 2 missing values
for mutation counts within the first 10 rows.</p>
<pre class="r"><code>metabric$mutation_count[1:10]</code></pre>
<pre><code>##  [1] NA  2  2  1  2  4  4 NA  1  5</code></pre>
<div class="rmdblock">
<p>
<strong>Missing values</strong> (<strong><code>NA</code></strong>)
</p>
<p>
Missing values in R are represented as <strong><code>NA</code></strong>,
which stands for ‘not available’.
</p>
</div>
<p>The <code>summary()</code> function reported that there are 152
missing mutation counts. <code>summary()</code> reports the mean and
median values for the remaining observations, i.e. those for which a
mutation count <em>is</em> available. Not all functions do this
though.</p>
<pre class="r"><code>mean(metabric$mutation_count)</code></pre>
<pre><code>## [1] NA</code></pre>
<p>The <code>mean()</code> function, for example, takes a different
approach in how it treats missing values. It returns <code>NA</code>,
i.e. a missing value, because it takes the view that it cannot compute a
mean for a set of values where some are not available. You can specify
the argument <code>na.rm = TRUE</code> when calling <code>mean()</code>
to instruct it to ignore the missing values and calculate the mean of
the remaining values.</p>
<pre class="r"><code>mean(metabric$mutation_count, na.rm = TRUE)</code></pre>
<pre><code>## [1] 5.578702</code></pre>
<p>There are many such functions in R that would be similarly affected
by missing values and which offer the <code>na.rm</code> argument.</p>
<p>You can test for missing values using the <code>is.na()</code>
function.</p>
<pre class="r"><code>first_ten_mutation_counts &lt;- metabric$mutation_count[1:10]
is.na(first_ten_mutation_counts)</code></pre>
<pre><code>##  [1]  TRUE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE</code></pre>
<p>One final handy tip: summing the result of <code>is.na()</code> will
count the number of missing values in a vector.</p>
<pre class="r"><code>sum(is.na(metabric$mutation_count))</code></pre>
<pre><code>## [1] 152</code></pre>
</div>
<div id="data-manipulation-with-dplyr" class="section level1">
<h1>Data manipulation with dplyr</h1>
<p><strong>dplyr</strong> is one of the packages that gets loaded as
part of the tidyverse.</p>
<pre class="r"><code>library(tidyverse)</code></pre>
<p>dplyr is the Swiss army knife in the tidyverse, providing many useful
functions for manipulating tabular data in data frames or tibbles. We’re
going to look at the key functions for filtering our data, modifying the
contents and computing summary statistics.</p>
<p>We’ll also introduce the pipe operator,
<strong><code>%&gt;%</code></strong>, for chaining operations together
into mini workflows in a way that makes for more readable and
maintainable code.</p>
<p>We’ll again use the METABRIC data set to illustrate how these
operations work.</p>
<pre class="r"><code>metabric &lt;- read_csv(&quot;data/metabric_clinical_and_expression_data.csv&quot;)
metabric</code></pre>
<pre><code>## # A tibble: 1,904 × 32
##    Patient_ID Cohort Age_at_di…¹ Survi…² Survi…³ Vital…⁴ Chemo…⁵ Radio…⁶ Tumou…⁷
##    &lt;chr&gt;       &lt;dbl&gt;       &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;     &lt;dbl&gt;
##  1 MB-0000         1        75.6   140.  LIVING  Living  NO      YES          22
##  2 MB-0002         1        43.2    84.6 LIVING  Living  NO      YES          10
##  3 MB-0005         1        48.9   164.  DECEAS… Died o… YES     NO           15
##  4 MB-0006         1        47.7   165.  LIVING  Living  YES     YES          25
##  5 MB-0008         1        77.0    41.4 DECEAS… Died o… YES     YES          40
##  6 MB-0010         1        78.8     7.8 DECEAS… Died o… NO      YES          31
##  7 MB-0014         1        56.4   164.  LIVING  Living  YES     YES          10
##  8 MB-0022         1        89.1    99.5 DECEAS… Died o… NO      YES          29
##  9 MB-0028         1        86.4    36.6 DECEAS… Died o… NO      YES          16
## 10 MB-0035         1        84.2    36.3 DECEAS… Died o… NO      NO           28
## # … with 1,894 more rows, 23 more variables: Tumour_stage &lt;dbl&gt;,
## #   Neoplasm_histologic_grade &lt;dbl&gt;, Lymph_nodes_examined_positive &lt;dbl&gt;,
## #   Lymph_node_status &lt;dbl&gt;, Cancer_type &lt;chr&gt;, ER_status &lt;chr&gt;,
## #   PR_status &lt;chr&gt;, HER2_status &lt;chr&gt;, HER2_status_measured_by_SNP6 &lt;chr&gt;,
## #   PAM50 &lt;chr&gt;, `3-gene_classifier` &lt;chr&gt;, Nottingham_prognostic_index &lt;dbl&gt;,
## #   Cellularity &lt;chr&gt;, Integrative_cluster &lt;chr&gt;, Mutation_count &lt;dbl&gt;,
## #   ESR1 &lt;dbl&gt;, ERBB2 &lt;dbl&gt;, PGR &lt;dbl&gt;, TP53 &lt;dbl&gt;, PIK3CA &lt;dbl&gt;, …</code></pre>
</div>
<div id="data-semantics" class="section level1">
<h1>Data semantics</h1>
<p>We use the terms ‘observation’ and ‘variable’ a lot in this course.
As a reminder from week 2, when we talk about an observation, we’re
talking about a set of values measured for the same unit or thing,
e.g. a person or a date, and when we talk about a variable we are really
talking about the attribute that we are measuring or recording,
e.g. height, temperature or expression value.</p>
<p><strong>Observations are represented as rows</strong> in our data
frames or tibbles, while the <strong>columns correspond to
variables</strong>.</p>
<div class="rmdblock">
<p>
From <strong>“Tidy Data”</strong> by <strong>Hadley Wickham</strong>, <a
href="https://www.jstatsoft.org/index.php/jss/article/view/v059i10/v59i10.pdf"><em>The
Journal of Statistical Software</em>, vol. 59, 2014</a>.
</p>
<p>
A data set is a collection of <strong>values</strong>, usually either
numbers (if quantitative) or character strings (if qualitative). Values
are organised in two ways. Every value belongs to a variable and an
observation.
</p>
<p>
A <strong>variable</strong> contains all values that measure the same
underlying attribute (like height, temperature, duration) across units.
</p>
<p>
An <strong>observation</strong> contains all values measured on the same
unit (like a person or a day) across attributes.
</p>
</div>
</div>
<div id="dplyr-verbs" class="section level1">
<h1>dplyr verbs</h1>
<p>We will be looking at the following key dplyr functions in this
course:</p>
<ul>
<li><strong><code>filter()</code></strong> for filtering rows</li>
<li><strong><code>select()</code></strong> for selecting columns</li>
<li><strong><code>arrange()</code></strong> for sorting rows</li>
<li><strong><code>mutate()</code></strong> for modifying columns or
creating new ones</li>
<li><strong><code>summarise()</code></strong> for computing summary
values</li>
</ul>
<p>In looking at each of these in turn, we’ll be applying these to the
entire data set. It is possible to combine these with the
<code>group_by()</code> function to instead operate on separate groups
within our data set but this is something we’ll cover in detail
later.</p>
<p>The dplyr operations are commonly referred to as “verbs” in a data
manipulation grammar. These verbs have a common syntax and work together
in a consistent and uniform manner. They all have the following shared
behaviours:</p>
<ul>
<li><p>The first argument in each function is a data frame (or
tibble)</p></li>
<li><p>Any additional arguments describe what operation to perform on
the data frame</p></li>
<li><p>Variable names, i.e. column names, are referred to without using
quotes</p></li>
<li><p>The result of an operation is a new data frame</p></li>
</ul>
</div>
<div id="filtering-rows-with-filter" class="section level1">
<h1>Filtering rows with <code>filter()</code></h1>
<p>The <strong><code>filter</code></strong> verb allows you to choose
rows from a data frame that match some specified criteria. The criteria
are based on values of variables and can make use of comparison
operators such as <code>==</code>, <code>&gt;</code>, <code>&lt;</code>
and <code>!=</code>.</p>
<p>For example, to filter the METABRIC data set so that it only contains
observations for those patients who died of breast cancer we can use
<code>filter()</code> as follows.</p>
<pre class="r"><code>deceased &lt;- filter(metabric, Vital_status == &quot;Died of Disease&quot;)
deceased</code></pre>
<pre><code>## # A tibble: 622 × 32
##    Patient_ID Cohort Age_at_di…¹ Survi…² Survi…³ Vital…⁴ Chemo…⁵ Radio…⁶ Tumou…⁷
##    &lt;chr&gt;       &lt;dbl&gt;       &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;     &lt;dbl&gt;
##  1 MB-0005         1        48.9  164.   DECEAS… Died o… YES     NO           15
##  2 MB-0008         1        77.0   41.4  DECEAS… Died o… YES     YES          40
##  3 MB-0010         1        78.8    7.8  DECEAS… Died o… NO      YES          31
##  4 MB-0035         1        84.2   36.3  DECEAS… Died o… NO      NO           28
##  5 MB-0036         1        85.5  132.   DECEAS… Died o… NO      YES          22
##  6 MB-0079         1        50.4   28.5  DECEAS… Died o… YES     YES          40
##  7 MB-0083         1        64.8   86.1  DECEAS… Died o… NO      YES          13
##  8 MB-0100         1        68.7    8.07 DECEAS… Died o… YES     YES          39
##  9 MB-0102         1        51.4  141.   DECEAS… Died o… YES     YES          40
## 10 MB-0108         1        43.2   42.7  DECEAS… Died o… YES     YES          18
## # … with 612 more rows, 23 more variables: Tumour_stage &lt;dbl&gt;,
## #   Neoplasm_histologic_grade &lt;dbl&gt;, Lymph_nodes_examined_positive &lt;dbl&gt;,
## #   Lymph_node_status &lt;dbl&gt;, Cancer_type &lt;chr&gt;, ER_status &lt;chr&gt;,
## #   PR_status &lt;chr&gt;, HER2_status &lt;chr&gt;, HER2_status_measured_by_SNP6 &lt;chr&gt;,
## #   PAM50 &lt;chr&gt;, `3-gene_classifier` &lt;chr&gt;, Nottingham_prognostic_index &lt;dbl&gt;,
## #   Cellularity &lt;chr&gt;, Integrative_cluster &lt;chr&gt;, Mutation_count &lt;dbl&gt;,
## #   ESR1 &lt;dbl&gt;, ERBB2 &lt;dbl&gt;, PGR &lt;dbl&gt;, TP53 &lt;dbl&gt;, PIK3CA &lt;dbl&gt;, …</code></pre>
<p>Remember that the <strong><code>==</code></strong> operator tests for
equality, i.e. is the value for <code>Vital_status</code> for each
patient (observation) equal to “Died of Disease”.</p>
<p>This filtering operation is equivalent to subsetting the rows based
on a logical vector resulting from our comparison of vital status values
with “Died of Disease”.</p>
<pre class="r"><code>filter(metabric, Vital_status == &quot;Died of Disease&quot;)
# is equivalent to
metabric[metabric$Vital_status == &quot;Died of Disease&quot;, ]</code></pre>
<p>Both achieve the same result but the dplyr filter method is arguably
a little easier to read. We haven’t had to write <code>metabric</code>
twice for one thing; we just referred to the variable name as it is,
unquoted and without any fuss.</p>
<p>Let’s have a look at the various categories in the
<code>Vital_status</code> variable.</p>
<pre class="r"><code>table(metabric$Vital_status)</code></pre>
<pre><code>## 
##      Died of Disease Died of Other Causes               Living 
##                  622                  480                  801</code></pre>
<p>We could use the <code>!=</code> comparison operator to select all
deceased patients regardless of whether they died of the disease or
other causes, by filtering for those that don’t have the value
“Living”.</p>
<pre class="r"><code>deceased &lt;- filter(metabric, Vital_status != &quot;Living&quot;)
nrow(deceased)</code></pre>
<pre><code>## [1] 1102</code></pre>
<p>Another way of doing this is to specify which classes we are
interested in and use the <strong><code>%in%</code></strong> operator to
test which observations (rows) contain those values.</p>
<pre class="r"><code>deceased &lt;- filter(metabric, Vital_status %in% c(&quot;Died of Disease&quot;, &quot;Died of Other Causes&quot;))
nrow(deceased)</code></pre>
<pre><code>## [1] 1102</code></pre>
<p>Another of the tidyverse packages,
<strong><code>stringr</code></strong>, contains a set of very useful
functions for operating on text or character strings. One such function,
<strong><code>str_starts()</code></strong> could be used to find all
<code>Vital_status</code> values that start with “Died”.</p>
<pre class="r"><code>deceased &lt;- filter(metabric, str_starts(Vital_status, &quot;Died&quot;))
nrow(deceased)</code></pre>
<pre><code>## [1] 1102</code></pre>
<p>Note that <code>str_starts()</code> returns a logical vector - this
is important since the filtering condition must evaluate to
<code>TRUE</code> or <code>FALSE</code> values for each row.</p>
<p>Unsurprisingly there is an equivalent function,
<code>str_ends()</code>, for matching the end of text (character) values
and <code>str_detect()</code> is another useful function that determines
whether values match a regular expression. Regular expressions are a
language for search patterns used frequently in computer programming and
really worth getting to grips with but sadly these are beyond the scope
of this course.</p>
<p>Filtering based on a logical variable is the most simple type of
filtering of all. We don’t have any logical variables in our METABRIC
data set so we’ll create one from the binary
<code>Survival_status</code> variable to use as an example.</p>
<pre class="r"><code># create a new logical variable called &#39;Deceased&#39;
metabric$Deceased &lt;- metabric$Survival_status == &quot;DECEASED&quot;
#
# filtering based on a logical variable - only selects TRUE values
deceased &lt;- filter(metabric, Deceased)
#
# only display those columns we&#39;re interested in
deceased[, c(&quot;Patient_ID&quot;, &quot;Survival_status&quot;, &quot;Vital_status&quot;, &quot;Deceased&quot;)]</code></pre>
<pre><code>## # A tibble: 1,103 × 4
##    Patient_ID Survival_status Vital_status         Deceased
##    &lt;chr&gt;      &lt;chr&gt;           &lt;chr&gt;                &lt;lgl&gt;   
##  1 MB-0005    DECEASED        Died of Disease      TRUE    
##  2 MB-0008    DECEASED        Died of Disease      TRUE    
##  3 MB-0010    DECEASED        Died of Disease      TRUE    
##  4 MB-0022    DECEASED        Died of Other Causes TRUE    
##  5 MB-0028    DECEASED        Died of Other Causes TRUE    
##  6 MB-0035    DECEASED        Died of Disease      TRUE    
##  7 MB-0036    DECEASED        Died of Disease      TRUE    
##  8 MB-0046    DECEASED        Died of Other Causes TRUE    
##  9 MB-0079    DECEASED        Died of Disease      TRUE    
## 10 MB-0083    DECEASED        Died of Disease      TRUE    
## # … with 1,093 more rows</code></pre>
<p>We can use the <strong><code>!</code></strong> operator to filter
those patients who are not deceased.</p>
<pre class="r"><code>filter(metabric, !Deceased)</code></pre>
<p>The eagle-eyed will have spotted that filtering on our newly created
Deceased logical variable gave a slightly different number of
observations (patients) who are considered to be deceased, compared with
the filtering operations shown above based on the
<code>Vital_status</code> variable. We get one extra row. This is
because we have a missing value for the vital status of one of the
patients. We can filter for this using the
<strong><code>is.na()</code></strong> function.</p>
<pre class="r"><code>missing_vital_status &lt;- filter(metabric, is.na(Vital_status))
missing_vital_status[, c(&quot;Patient_ID&quot;, &quot;Survival_status&quot;, &quot;Vital_status&quot;, &quot;Deceased&quot;)]</code></pre>
<pre><code>## # A tibble: 1 × 4
##   Patient_ID Survival_status Vital_status Deceased
##   &lt;chr&gt;      &lt;chr&gt;           &lt;chr&gt;        &lt;lgl&gt;   
## 1 MB-5130    DECEASED        &lt;NA&gt;         TRUE</code></pre>
<p><code>filter()</code> only retains rows where the condition if
<code>TRUE</code>; both <code>FALSE</code> and <code>NA</code> values
are filtered out.</p>
<p>We can apply more than one condition in our filtering operation, for
example the patients who were still alive at the time of the METABRIC
study and had survived for more than 10 years (120 months).</p>
<pre class="r"><code>filter(metabric, Survival_status == &quot;LIVING&quot;, Survival_time &gt; 120)</code></pre>
<pre><code>## # A tibble: 545 × 33
##    Patient_ID Cohort Age_at_di…¹ Survi…² Survi…³ Vital…⁴ Chemo…⁵ Radio…⁶ Tumou…⁷
##    &lt;chr&gt;       &lt;dbl&gt;       &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;     &lt;dbl&gt;
##  1 MB-0000         1        75.6    140. LIVING  Living  NO      YES          22
##  2 MB-0006         1        47.7    165. LIVING  Living  YES     YES          25
##  3 MB-0014         1        56.4    164. LIVING  Living  YES     YES          10
##  4 MB-0039         1        70.9    164. LIVING  Living  NO      YES          21
##  5 MB-0045         1        45.3    165. LIVING  Living  YES     YES          19
##  6 MB-0053         1        70.0    161. LIVING  Living  NO      YES          23
##  7 MB-0054         1        66.9    160. LIVING  Living  NO      YES          36
##  8 MB-0060         1        45.4    141. LIVING  Living  YES     YES          23
##  9 MB-0062         1        52.1    154. LIVING  Living  YES     YES          17
## 10 MB-0066         1        61.5    157. LIVING  Living  NO      YES          16
## # … with 535 more rows, 24 more variables: Tumour_stage &lt;dbl&gt;,
## #   Neoplasm_histologic_grade &lt;dbl&gt;, Lymph_nodes_examined_positive &lt;dbl&gt;,
## #   Lymph_node_status &lt;dbl&gt;, Cancer_type &lt;chr&gt;, ER_status &lt;chr&gt;,
## #   PR_status &lt;chr&gt;, HER2_status &lt;chr&gt;, HER2_status_measured_by_SNP6 &lt;chr&gt;,
## #   PAM50 &lt;chr&gt;, `3-gene_classifier` &lt;chr&gt;, Nottingham_prognostic_index &lt;dbl&gt;,
## #   Cellularity &lt;chr&gt;, Integrative_cluster &lt;chr&gt;, Mutation_count &lt;dbl&gt;,
## #   ESR1 &lt;dbl&gt;, ERBB2 &lt;dbl&gt;, PGR &lt;dbl&gt;, TP53 &lt;dbl&gt;, PIK3CA &lt;dbl&gt;, …</code></pre>
<p>The equivalent using R’s usual subsetting is slightly less
readable.</p>
<pre class="r"><code>metabric[metabric$Survival_status == &quot;LIVING&quot; &amp; metabric$Survival_time &gt; 120, ]</code></pre>
<p>We can add as many conditions as we like separating each with a
comma. Note that filtering using R subsetting gets more unreadable, and
more cumbersome to code, the more conditions you add.</p>
<p>Adding conditions in this way is equivalent to combining the
conditions using the <strong><code>&amp;</code></strong> (Boolean AND)
operator.</p>
<pre class="r"><code>filter(metabric, Survival_status == &quot;LIVING&quot;, Survival_time &gt; 120)</code></pre>
<p>Naturally we can also check when either of two conditions holds true
by using the <strong><code>|</code></strong> (Boolean OR) operator. And
we can build up more complicated filtering operations using both
<code>&amp;</code> and <code>|</code>. For example, let’s see which
patients have stage 3 or stage 4 tumours that are either estrogen
receptor (ER) positive or progesterone receptor (PR) positive.</p>
<pre class="r"><code>selected_patients &lt;- filter(metabric, Tumour_stage &gt;= 3, ER_status == &quot;Positive&quot; | PR_status == &quot;Positive&quot;)
nrow(selected_patients)</code></pre>
<pre><code>## [1] 79</code></pre>
<p>In this case, if you used <code>&amp;</code> in place of the comma
you’d need to be careful about the precedence of the <code>&amp;</code>
and <code>|</code> operators and use parentheses to make clear what you
intended.</p>
<pre class="r"><code>filter(metabric, Tumour_stage &gt;= 3 &amp; (ER_status == &quot;Positive&quot; | PR_status == &quot;Positive&quot;))</code></pre>
</div>
<div id="selecting-columns-with-select" class="section level1">
<h1>Selecting columns with <code>select()</code></h1>
<p>Another way of “slicing and dicing”” our tabular data set is to
select just the variables or columns we’re interested in. This can be
important particularly when the data set contains a very large number of
variables as is the case for the METABRIC data. Notice how when we print
the METABRIC data frame it is not possible to display all the columns;
we only get the first few and then a long list of the additional ones
that weren’t displayed.</p>
<p>Using the <code>$</code> operator is quite convenient for selecting a
single column and extracting the values as a vector. Selecting several
columns using the <code>[]</code> subsetting operator is a bit more
cumbersome. For example, in our look at filtering rows, we considered
two different variables in our data set that are concerned with the
living/deceased status of patients. When printing out the results we
selected just those columns along with the patient identifier and the
newly created <code>Deceased</code> column.</p>
<pre class="r"><code>deceased[, c(&quot;Patient_ID&quot;, &quot;Survival_status&quot;, &quot;Vital_status&quot;, &quot;Deceased&quot;)]</code></pre>
<pre><code>## # A tibble: 1,103 × 4
##    Patient_ID Survival_status Vital_status         Deceased
##    &lt;chr&gt;      &lt;chr&gt;           &lt;chr&gt;                &lt;lgl&gt;   
##  1 MB-0005    DECEASED        Died of Disease      TRUE    
##  2 MB-0008    DECEASED        Died of Disease      TRUE    
##  3 MB-0010    DECEASED        Died of Disease      TRUE    
##  4 MB-0022    DECEASED        Died of Other Causes TRUE    
##  5 MB-0028    DECEASED        Died of Other Causes TRUE    
##  6 MB-0035    DECEASED        Died of Disease      TRUE    
##  7 MB-0036    DECEASED        Died of Disease      TRUE    
##  8 MB-0046    DECEASED        Died of Other Causes TRUE    
##  9 MB-0079    DECEASED        Died of Disease      TRUE    
## 10 MB-0083    DECEASED        Died of Disease      TRUE    
## # … with 1,093 more rows</code></pre>
<p>The <strong><code>select()</code></strong> function from dplyr is
simpler.</p>
<pre class="r"><code>select(metabric, Patient_ID, Survival_status, Vital_status, Deceased)</code></pre>
<pre><code>## # A tibble: 1,904 × 4
##    Patient_ID Survival_status Vital_status         Deceased
##    &lt;chr&gt;      &lt;chr&gt;           &lt;chr&gt;                &lt;lgl&gt;   
##  1 MB-0000    LIVING          Living               FALSE   
##  2 MB-0002    LIVING          Living               FALSE   
##  3 MB-0005    DECEASED        Died of Disease      TRUE    
##  4 MB-0006    LIVING          Living               FALSE   
##  5 MB-0008    DECEASED        Died of Disease      TRUE    
##  6 MB-0010    DECEASED        Died of Disease      TRUE    
##  7 MB-0014    LIVING          Living               FALSE   
##  8 MB-0022    DECEASED        Died of Other Causes TRUE    
##  9 MB-0028    DECEASED        Died of Other Causes TRUE    
## 10 MB-0035    DECEASED        Died of Disease      TRUE    
## # … with 1,894 more rows</code></pre>
<p>Notice the similarities with the <code>filter()</code> function. The
first argument is the data frame we are operating on and the arguments
that follow on are specific to the operation in question, in this case,
the variables (columns) to select. Note that the variables do not need
to be put in quotes, and the returned value is another data frame, even
if only one column was selected.</p>
<p>We can alter the order of the variables (columns).</p>
<pre class="r"><code>select(metabric, Patient_ID, Vital_status, Survival_status, Deceased)</code></pre>
<pre><code>## # A tibble: 1,904 × 4
##    Patient_ID Vital_status         Survival_status Deceased
##    &lt;chr&gt;      &lt;chr&gt;                &lt;chr&gt;           &lt;lgl&gt;   
##  1 MB-0000    Living               LIVING          FALSE   
##  2 MB-0002    Living               LIVING          FALSE   
##  3 MB-0005    Died of Disease      DECEASED        TRUE    
##  4 MB-0006    Living               LIVING          FALSE   
##  5 MB-0008    Died of Disease      DECEASED        TRUE    
##  6 MB-0010    Died of Disease      DECEASED        TRUE    
##  7 MB-0014    Living               LIVING          FALSE   
##  8 MB-0022    Died of Other Causes DECEASED        TRUE    
##  9 MB-0028    Died of Other Causes DECEASED        TRUE    
## 10 MB-0035    Died of Disease      DECEASED        TRUE    
## # … with 1,894 more rows</code></pre>
<p>We can also select a range of columns using
<strong><code>:</code></strong>, e.g. to select the patient identifier
and all the columns between <code>Tumour_size</code> and
<code>Cancer_type</code> we could run the following
<code>select()</code> command.</p>
<pre class="r"><code>select(metabric, Patient_ID, Chemotherapy:Tumour_stage)</code></pre>
<pre><code>## # A tibble: 1,904 × 5
##    Patient_ID Chemotherapy Radiotherapy Tumour_size Tumour_stage
##    &lt;chr&gt;      &lt;chr&gt;        &lt;chr&gt;              &lt;dbl&gt;        &lt;dbl&gt;
##  1 MB-0000    NO           YES                   22            2
##  2 MB-0002    NO           YES                   10            1
##  3 MB-0005    YES          NO                    15            2
##  4 MB-0006    YES          YES                   25            2
##  5 MB-0008    YES          YES                   40            2
##  6 MB-0010    NO           YES                   31            4
##  7 MB-0014    YES          YES                   10            2
##  8 MB-0022    NO           YES                   29            2
##  9 MB-0028    NO           YES                   16            2
## 10 MB-0035    NO           NO                    28            2
## # … with 1,894 more rows</code></pre>
<p>The help page for <code>select</code> points to some special
functions that can be used within <code>select()</code>. We can find all
the columns, for example, that contain the term “status” using
<code>contains()</code>.</p>
<pre class="r"><code>select(metabric, contains(&quot;status&quot;))</code></pre>
<pre><code>## # A tibble: 1,904 × 7
##    Survival_status Vital_status         Lymph_…¹ ER_st…² PR_st…³ HER2_…⁴ HER2_…⁵
##    &lt;chr&gt;           &lt;chr&gt;                   &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;  
##  1 LIVING          Living                      3 Positi… Negati… Negati… NEUTRAL
##  2 LIVING          Living                      1 Positi… Positi… Negati… NEUTRAL
##  3 DECEASED        Died of Disease             2 Positi… Positi… Negati… NEUTRAL
##  4 LIVING          Living                      2 Positi… Positi… Negati… NEUTRAL
##  5 DECEASED        Died of Disease             3 Positi… Positi… Negati… NEUTRAL
##  6 DECEASED        Died of Disease             1 Positi… Positi… Negati… NEUTRAL
##  7 LIVING          Living                      2 Positi… Positi… Negati… LOSS   
##  8 DECEASED        Died of Other Causes        2 Positi… Negati… Negati… NEUTRAL
##  9 DECEASED        Died of Other Causes        2 Positi… Negati… Negati… GAIN   
## 10 DECEASED        Died of Disease             1 Positi… Negati… Negati… LOSS   
## # … with 1,894 more rows, and abbreviated variable names ¹​Lymph_node_status,
## #   ²​ER_status, ³​PR_status, ⁴​HER2_status, ⁵​HER2_status_measured_by_SNP6</code></pre>
<p>If we only wanted those ending with “status” we could use
<code>ends_with()</code> instead.</p>
<pre class="r"><code>select(metabric, ends_with(&quot;status&quot;))</code></pre>
<pre><code>## # A tibble: 1,904 × 6
##    Survival_status Vital_status         Lymph_node_sta…¹ ER_st…² PR_st…³ HER2_…⁴
##    &lt;chr&gt;           &lt;chr&gt;                           &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;  
##  1 LIVING          Living                              3 Positi… Negati… Negati…
##  2 LIVING          Living                              1 Positi… Positi… Negati…
##  3 DECEASED        Died of Disease                     2 Positi… Positi… Negati…
##  4 LIVING          Living                              2 Positi… Positi… Negati…
##  5 DECEASED        Died of Disease                     3 Positi… Positi… Negati…
##  6 DECEASED        Died of Disease                     1 Positi… Positi… Negati…
##  7 LIVING          Living                              2 Positi… Positi… Negati…
##  8 DECEASED        Died of Other Causes                2 Positi… Negati… Negati…
##  9 DECEASED        Died of Other Causes                2 Positi… Negati… Negati…
## 10 DECEASED        Died of Disease                     1 Positi… Negati… Negati…
## # … with 1,894 more rows, and abbreviated variable names ¹​Lymph_node_status,
## #   ²​ER_status, ³​PR_status, ⁴​HER2_status</code></pre>
<p>We can also select those columns we’re not interested in and that
shouldn’t be included by prefixing the columns with
<strong><code>-</code></strong>.</p>
<pre class="r"><code>select(metabric, -Cohort)</code></pre>
<pre><code>## # A tibble: 1,904 × 32
##    Patient_ID Age_at_d…¹ Survi…² Survi…³ Vital…⁴ Chemo…⁵ Radio…⁶ Tumou…⁷ Tumou…⁸
##    &lt;chr&gt;           &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;     &lt;dbl&gt;   &lt;dbl&gt;
##  1 MB-0000          75.6   140.  LIVING  Living  NO      YES          22       2
##  2 MB-0002          43.2    84.6 LIVING  Living  NO      YES          10       1
##  3 MB-0005          48.9   164.  DECEAS… Died o… YES     NO           15       2
##  4 MB-0006          47.7   165.  LIVING  Living  YES     YES          25       2
##  5 MB-0008          77.0    41.4 DECEAS… Died o… YES     YES          40       2
##  6 MB-0010          78.8     7.8 DECEAS… Died o… NO      YES          31       4
##  7 MB-0014          56.4   164.  LIVING  Living  YES     YES          10       2
##  8 MB-0022          89.1    99.5 DECEAS… Died o… NO      YES          29       2
##  9 MB-0028          86.4    36.6 DECEAS… Died o… NO      YES          16       2
## 10 MB-0035          84.2    36.3 DECEAS… Died o… NO      NO           28       2
## # … with 1,894 more rows, 23 more variables: Neoplasm_histologic_grade &lt;dbl&gt;,
## #   Lymph_nodes_examined_positive &lt;dbl&gt;, Lymph_node_status &lt;dbl&gt;,
## #   Cancer_type &lt;chr&gt;, ER_status &lt;chr&gt;, PR_status &lt;chr&gt;, HER2_status &lt;chr&gt;,
## #   HER2_status_measured_by_SNP6 &lt;chr&gt;, PAM50 &lt;chr&gt;, `3-gene_classifier` &lt;chr&gt;,
## #   Nottingham_prognostic_index &lt;dbl&gt;, Cellularity &lt;chr&gt;,
## #   Integrative_cluster &lt;chr&gt;, Mutation_count &lt;dbl&gt;, ESR1 &lt;dbl&gt;, ERBB2 &lt;dbl&gt;,
## #   PGR &lt;dbl&gt;, TP53 &lt;dbl&gt;, PIK3CA &lt;dbl&gt;, GATA3 &lt;dbl&gt;, FOXA1 &lt;dbl&gt;, …</code></pre>
<p>You can use a combination of explicit naming, ranges, helper
functions and negation to select the columns of interest.</p>
<pre class="r"><code>selected_columns &lt;- select(metabric, Patient_ID, starts_with(&quot;Tumour_&quot;), `3-gene_classifier`:Integrative_cluster, -Cellularity)
selected_columns</code></pre>
<pre><code>## # A tibble: 1,904 × 6
##    Patient_ID Tumour_size Tumour_stage `3-gene_classifier`   Nottingha…¹ Integ…²
##    &lt;chr&gt;            &lt;dbl&gt;        &lt;dbl&gt; &lt;chr&gt;                       &lt;dbl&gt; &lt;chr&gt;  
##  1 MB-0000             22            2 ER-/HER2-                    6.04 4ER+   
##  2 MB-0002             10            1 ER+/HER2- High Prolif        4.02 4ER+   
##  3 MB-0005             15            2 &lt;NA&gt;                         4.03 3      
##  4 MB-0006             25            2 &lt;NA&gt;                         4.05 9      
##  5 MB-0008             40            2 ER+/HER2- High Prolif        6.08 9      
##  6 MB-0010             31            4 ER+/HER2- High Prolif        4.06 7      
##  7 MB-0014             10            2 &lt;NA&gt;                         4.02 3      
##  8 MB-0022             29            2 &lt;NA&gt;                         4.06 3      
##  9 MB-0028             16            2 ER+/HER2- High Prolif        5.03 9      
## 10 MB-0035             28            2 ER+/HER2- High Prolif        3.06 3      
## # … with 1,894 more rows, and abbreviated variable names
## #   ¹​Nottingham_prognostic_index, ²​Integrative_cluster</code></pre>
<p>You can also use <code>select()</code> to completely reorder the
columns so they’re in the order of your choosing. The
<code>everything()</code> helper function is useful in this context,
particularly if what you’re really interested in is bringing one or more
columns to the left hand side and then including everything else
afterwards in whatever order they were already in.</p>
<pre class="r"><code>select(metabric, Patient_ID, Survival_status, Tumour_stage, everything())</code></pre>
<pre><code>## # A tibble: 1,904 × 33
##    Patient_ID Survival_…¹ Tumou…² Cohort Age_a…³ Survi…⁴ Vital…⁵ Chemo…⁶ Radio…⁷
##    &lt;chr&gt;      &lt;chr&gt;         &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;  
##  1 MB-0000    LIVING            2      1    75.6   140.  Living  NO      YES    
##  2 MB-0002    LIVING            1      1    43.2    84.6 Living  NO      YES    
##  3 MB-0005    DECEASED          2      1    48.9   164.  Died o… YES     NO     
##  4 MB-0006    LIVING            2      1    47.7   165.  Living  YES     YES    
##  5 MB-0008    DECEASED          2      1    77.0    41.4 Died o… YES     YES    
##  6 MB-0010    DECEASED          4      1    78.8     7.8 Died o… NO      YES    
##  7 MB-0014    LIVING            2      1    56.4   164.  Living  YES     YES    
##  8 MB-0022    DECEASED          2      1    89.1    99.5 Died o… NO      YES    
##  9 MB-0028    DECEASED          2      1    86.4    36.6 Died o… NO      YES    
## 10 MB-0035    DECEASED          2      1    84.2    36.3 Died o… NO      NO     
## # … with 1,894 more rows, 24 more variables: Tumour_size &lt;dbl&gt;,
## #   Neoplasm_histologic_grade &lt;dbl&gt;, Lymph_nodes_examined_positive &lt;dbl&gt;,
## #   Lymph_node_status &lt;dbl&gt;, Cancer_type &lt;chr&gt;, ER_status &lt;chr&gt;,
## #   PR_status &lt;chr&gt;, HER2_status &lt;chr&gt;, HER2_status_measured_by_SNP6 &lt;chr&gt;,
## #   PAM50 &lt;chr&gt;, `3-gene_classifier` &lt;chr&gt;, Nottingham_prognostic_index &lt;dbl&gt;,
## #   Cellularity &lt;chr&gt;, Integrative_cluster &lt;chr&gt;, Mutation_count &lt;dbl&gt;,
## #   ESR1 &lt;dbl&gt;, ERBB2 &lt;dbl&gt;, PGR &lt;dbl&gt;, TP53 &lt;dbl&gt;, PIK3CA &lt;dbl&gt;, …</code></pre>
<p>Finally, columns can be renamed as part of the selection process.</p>
<pre class="r"><code>select(metabric, Patient_ID, Classification = `3-gene_classifier`, NPI = Nottingham_prognostic_index)</code></pre>
<pre><code>## # A tibble: 1,904 × 3
##    Patient_ID Classification          NPI
##    &lt;chr&gt;      &lt;chr&gt;                 &lt;dbl&gt;
##  1 MB-0000    ER-/HER2-              6.04
##  2 MB-0002    ER+/HER2- High Prolif  4.02
##  3 MB-0005    &lt;NA&gt;                   4.03
##  4 MB-0006    &lt;NA&gt;                   4.05
##  5 MB-0008    ER+/HER2- High Prolif  6.08
##  6 MB-0010    ER+/HER2- High Prolif  4.06
##  7 MB-0014    &lt;NA&gt;                   4.02
##  8 MB-0022    &lt;NA&gt;                   4.06
##  9 MB-0028    ER+/HER2- High Prolif  5.03
## 10 MB-0035    ER+/HER2- High Prolif  3.06
## # … with 1,894 more rows</code></pre>
<p>Note that dplyr provides the <code>rename()</code> function for when
we only want to rename a column and not select a subset of columns.</p>
</div>
<div id="sorting-using-arrange" class="section level1">
<h1>Sorting using <code>arrange()</code></h1>
<p>It is sometimes quite useful to sort our data frame based on the
values in one or more of the columns, particularly when displaying the
contents or saving them to a file. The <code>arrange()</code> function
in dplyr provides this sorting capability.</p>
<p>For example, we can sort the METABRIC patients into order of
increasing survival time.</p>
<pre class="r"><code>arrange(metabric, Survival_time)</code></pre>
<pre><code>## # A tibble: 1,904 × 33
##    Patient_ID Cohort Age_at_di…¹ Survi…² Survi…³ Vital…⁴ Chemo…⁵ Radio…⁶ Tumou…⁷
##    &lt;chr&gt;       &lt;dbl&gt;       &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;     &lt;dbl&gt;
##  1 MB-0284         1        51.4   0     LIVING  Living  NO      NO           NA
##  2 MB-6229         5        75.3   0.1   DECEAS… Died o… NO      NO           15
##  3 MB-0627         1        54.1   0.767 LIVING  Living  NO      NO           20
##  4 MB-0880         1        73.6   1.23  LIVING  Living  NO      NO           30
##  5 MB-0125         1        74.0   1.27  LIVING  Living  NO      NO            9
##  6 MB-0374         1        34.7   1.43  LIVING  Living  NO      NO           25
##  7 MB-0148         1        53.2   1.77  LIVING  Living  NO      NO           19
##  8 MB-5525         3        63.2   2     LIVING  Living  NO      YES          28
##  9 MB-6092         5        80.6   2.3   DECEAS… Died o… NO      NO           21
## 10 MB-0117         1        60.1   2.4   LIVING  Living  NO      YES          23
## # … with 1,894 more rows, 24 more variables: Tumour_stage &lt;dbl&gt;,
## #   Neoplasm_histologic_grade &lt;dbl&gt;, Lymph_nodes_examined_positive &lt;dbl&gt;,
## #   Lymph_node_status &lt;dbl&gt;, Cancer_type &lt;chr&gt;, ER_status &lt;chr&gt;,
## #   PR_status &lt;chr&gt;, HER2_status &lt;chr&gt;, HER2_status_measured_by_SNP6 &lt;chr&gt;,
## #   PAM50 &lt;chr&gt;, `3-gene_classifier` &lt;chr&gt;, Nottingham_prognostic_index &lt;dbl&gt;,
## #   Cellularity &lt;chr&gt;, Integrative_cluster &lt;chr&gt;, Mutation_count &lt;dbl&gt;,
## #   ESR1 &lt;dbl&gt;, ERBB2 &lt;dbl&gt;, PGR &lt;dbl&gt;, TP53 &lt;dbl&gt;, PIK3CA &lt;dbl&gt;, …</code></pre>
<p>Or we might be more interested in the patients that survived the
longest so would need the order to be of decreasing survival time. For
that we can use the helper function <code>desc()</code> that works
specifically with <code>arrange()</code>.</p>
<pre class="r"><code>arrange(metabric, desc(Survival_time))</code></pre>
<pre><code>## # A tibble: 1,904 × 33
##    Patient_ID Cohort Age_at_di…¹ Survi…² Survi…³ Vital…⁴ Chemo…⁵ Radio…⁶ Tumou…⁷
##    &lt;chr&gt;       &lt;dbl&gt;       &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;     &lt;dbl&gt;
##  1 MB-4189         3        61.0    355. DECEAS… Died o… NO      YES          NA
##  2 MB-4079         3        63.2    351  DECEAS… Died o… NO      YES          22
##  3 MB-0270         1        30.0    337. LIVING  Living  NO      NO           14
##  4 MB-4235         3        67.5    336. DECEAS… Died o… NO      NO           20
##  5 MB-4292         3        58.8    336. DECEAS… Died o… NO      YES          25
##  6 MB-4212         3        45.5    330. LIVING  Living  NO      NO           20
##  7 MB-4548         3        50.4    323. LIVING  Living  YES     YES           1
##  8 MB-4633         3        67.0    318. LIVING  Living  NO      YES          25
##  9 MB-4332         3        34.4    308. LIVING  Living  NO      YES          20
## 10 MB-4418         3        56.1    308. LIVING  Living  NO      YES          20
## # … with 1,894 more rows, 24 more variables: Tumour_stage &lt;dbl&gt;,
## #   Neoplasm_histologic_grade &lt;dbl&gt;, Lymph_nodes_examined_positive &lt;dbl&gt;,
## #   Lymph_node_status &lt;dbl&gt;, Cancer_type &lt;chr&gt;, ER_status &lt;chr&gt;,
## #   PR_status &lt;chr&gt;, HER2_status &lt;chr&gt;, HER2_status_measured_by_SNP6 &lt;chr&gt;,
## #   PAM50 &lt;chr&gt;, `3-gene_classifier` &lt;chr&gt;, Nottingham_prognostic_index &lt;dbl&gt;,
## #   Cellularity &lt;chr&gt;, Integrative_cluster &lt;chr&gt;, Mutation_count &lt;dbl&gt;,
## #   ESR1 &lt;dbl&gt;, ERBB2 &lt;dbl&gt;, PGR &lt;dbl&gt;, TP53 &lt;dbl&gt;, PIK3CA &lt;dbl&gt;, …</code></pre>
<p>We can sort by more than one variable by adding more variable
arguments to <code>arrange()</code>.</p>
<pre class="r"><code>arrange(metabric, Tumour_stage, Nottingham_prognostic_index)</code></pre>
<pre><code>## # A tibble: 1,904 × 33
##    Patient_ID Cohort Age_at_di…¹ Survi…² Survi…³ Vital…⁴ Chemo…⁵ Radio…⁶ Tumou…⁷
##    &lt;chr&gt;       &lt;dbl&gt;       &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;     &lt;dbl&gt;
##  1 MB-0472         1        59.3    27.4 LIVING  Living  NO      YES          11
##  2 MB-0230         1        42.0   200.  LIVING  Living  NO      NO           35
##  3 MB-0522         1        54.1     2.5 LIVING  Living  NO      NO           NA
##  4 MB-0573         1        76.2   163.  DECEAS… Died o… NO      YES          65
##  5 MB-5186         3        42.8   203.  LIVING  Living  NO      YES          10
##  6 MB-0260         1        78.3   170.  DECEAS… Died o… NO      YES          12
##  7 MB-0245         1        54.8   165.  LIVING  Living  NO      NO           14
##  8 MB-2821         2        57.4   136.  DECEAS… Died o… NO      YES          16
##  9 MB-4681         3        60.7   252.  DECEAS… Died o… NO      NO           17
## 10 MB-0318         1        58.0   169.  LIVING  Living  NO      NO           18
## # … with 1,894 more rows, 24 more variables: Tumour_stage &lt;dbl&gt;,
## #   Neoplasm_histologic_grade &lt;dbl&gt;, Lymph_nodes_examined_positive &lt;dbl&gt;,
## #   Lymph_node_status &lt;dbl&gt;, Cancer_type &lt;chr&gt;, ER_status &lt;chr&gt;,
## #   PR_status &lt;chr&gt;, HER2_status &lt;chr&gt;, HER2_status_measured_by_SNP6 &lt;chr&gt;,
## #   PAM50 &lt;chr&gt;, `3-gene_classifier` &lt;chr&gt;, Nottingham_prognostic_index &lt;dbl&gt;,
## #   Cellularity &lt;chr&gt;, Integrative_cluster &lt;chr&gt;, Mutation_count &lt;dbl&gt;,
## #   ESR1 &lt;dbl&gt;, ERBB2 &lt;dbl&gt;, PGR &lt;dbl&gt;, TP53 &lt;dbl&gt;, PIK3CA &lt;dbl&gt;, …</code></pre>
<p>Here we’ve sorted first by tumour stage from lowest to highest value
and then by the Nottingham prognostic index for when there are ties,
i.e. where the tumour stage is the same.</p>
<p>Sorting is most commonly used in workflows as one of the last steps
before printing out a data frame or writing out the table to a file.</p>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p>In this session we have covered the following concepts:</p>
<ul>
<li>Create an R Project</li>
<li>How to install and use packages that extend R’s basic
functionality</li>
<li>How to read tabular data into R</li>
<li>The tibble, a superior version of the data frame</li>
<li>Missing values and how R handles these</li>
<li>Filtering rows in a data frame based on their values</li>
<li>Selecting and reordering of columns</li>
<li>Sorting data</li>
</ul>
<div id="exercises" class="section level2">
<h2>Exercises</h2>
<p><strong>1. How many samples in the METABRIC dataset have high
cellularity and positive ER and HER2 status?</strong></p>
<p>After reading the data first subset it to retain only cellularity,
er_status and her2_status. Thereafter, get rid of rows with missing
values in any of the columns.</p>
<pre class="r"><code>#library(tidyverse)
#metabric &lt;- read_tsv(&quot;data/brca_metabric_clinical_data.tsv&quot;)</code></pre>
<p><strong>2. Investigate the subset of long-surviving breast cancer
patients that didn’t receive chemo or radiotherapy</strong></p>
<p>First obtain the subset of patients that received neither
chemotherapy or radiotherapy and survived for more than 20 years.</p>
<pre class="r"><code>#library(tidyverse)
#metabric &lt;- read_csv(&quot;metabric_clinical_and_expression_data.csv&quot;)</code></pre>
<p>Now look at the breakdown of these patients in terms of ER status.
Count the numbers of ER positive and ER negative patients in this
subset. Calculate the proportion that are ER positive.</p>
<p>What does this tell us? Calculate the proportion of ER positive
patients in the whole cohort by way of a comparison.</p>
<p><strong>3. Read into R one of your own Excel spreadsheets and explore
the data using <code>summary()</code>, <code>table()</code> and/or
<code>plot()</code></strong></p>
<p>We haven’t yet showed how to read Excel spreadsheets into R. There
are more than package providing this capability but for this exercise
use the <code>readxl</code> package from the tidyverse.</p>
<p><code>readxl</code> should have been installed as part of the
tidyverse but if you find that’s not the case, you will have to install
it separately.</p>
<p><code>readxl</code> is not loaded as one of the core tidyverse
packages when you run <code>library(tidyverse)</code> so you’ll have to
load this separately.</p>
<p>If you don’t have an Excel spreadsheet to hand, you can use the
clinical annotation data for the NCI-60 cell line data set (<a
href="https://bioinformatics-core-shared-training.github.io/Bitesize-R/data/cellline_nci60_clinical_data.xlsx"
class="uri">https://bioinformatics-core-shared-training.github.io/Bitesize-R/data/cellline_nci60_clinical_data.xlsx</a>).</p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
