<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Ashley D. Sawle" />


<title>Exploring the Metabric data set</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img style="width: 120px; float: left; margin-top: 0px; margin-right: 20px;" src="images/CRUK_CI_logo.png"/>R for Biologists&nbsp;&nbsp;&nbsp;</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Contents
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="week1.html">Week 1 - Introduction to R</a>
    </li>
    <li>
      <a href="week2.html">Week 2 - Introduction to R data structures</a>
    </li>
    <li>
      <a href="week3.html">Week 3 - Working with data in R</a>
    </li>
    <li>
      <a href="week4.html">Week 4 - Data visualization with ggplot2</a>
    </li>
    <li>
      <a href="week5.html">Week 5 - Data manipulation using dplyr</a>
    </li>
    <li>
      <a href="week6.html">Week 6 - Grouping and combining data</a>
    </li>
    <li>
      <a href="week8.html">Week 8 - Restructuring data for analysis</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/bioinformatics-core-shared-training/Bitesize-R">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Exploring the Metabric data set</h1>
<h3 class="subtitle">An example of using R Markdown</h3>
<h4 class="author">Ashley D. Sawle</h4>
<h4 class="date">24th April 2025</h4>

</div>


<pre><code>Code chunk: &quot;setup&quot;</code></pre>
<div id="preamble" class="section level2">
<h2>Preamble</h2>
<p>The purpose of this document is to show how R Markdown can be used to
generate a report in HTML format.</p>
</div>
<div id="the-data" class="section level2">
<h2>The data</h2>
<p>We’ll be using an extended version of the METABRIC data set in which
columns have been added for the mRNA expression values for selected
genes, including estrogen receptor alpha (<strong>ESR1</strong>),
progesterone receptor (<strong>PGR</strong>), <strong>GATA3</strong> and
<strong>FOXA1</strong>.</p>
<p>The METABRIC study characterized the genomic mutations and gene
expression profiles for over 2000 primary breast tumours.</p>
<p><a href="https://pubmed.ncbi.nlm.nih.gov/22522925">Curtis <em>et
al.</em>, Nature 486:346-52, 2012</a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/pubmed/27161491">Pereira <em>et
al.</em>, Nature Communications 7:11479, 2016</a></p>
<p>Both the clinical data and the gene expression values were downloaded
from <a
href="https://www.cbioportal.org/study/summary?id=brca_metabric">cBioPortal</a>.</p>
<pre class="r"><code>metabric &lt;- read_csv(&quot;data/metabric_clinical_and_expression_data.csv&quot;) %&gt;%
    select(Patient_ID,
           ER_status,
           HER2_status,
           `3-gene_classifier`,
           PAM50, ESR1, ERBB2, PGR, TP53, PIK3CA, GATA3, FOXA1, MLPH)</code></pre>
<p>The table below shows the first few rows of the data for the first 7
columns.</p>
<pre><code>Code chunk: top few rows of table. Note the alignment in the columns.</code></pre>
<table class="table table-striped table-hover" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
Patient_ID
</th>
<th style="text-align:center;">
ER_status
</th>
<th style="text-align:center;">
HER2_status
</th>
<th style="text-align:center;">
3-gene_classifier
</th>
<th style="text-align:center;">
PAM50
</th>
<th style="text-align:right;">
ESR1
</th>
<th style="text-align:right;">
ERBB2
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;font-weight: bold;border-right:1px solid;">
MB-0000
</td>
<td style="text-align:center;">
Positive
</td>
<td style="text-align:center;">
Negative
</td>
<td style="text-align:center;">
ER-/HER2-
</td>
<td style="text-align:center;">
claudin-low
</td>
<td style="text-align:right;">
8.93
</td>
<td style="text-align:right;">
9.33
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;border-right:1px solid;">
MB-0002
</td>
<td style="text-align:center;">
Positive
</td>
<td style="text-align:center;">
Negative
</td>
<td style="text-align:center;">
ER+/HER2- High Prolif
</td>
<td style="text-align:center;">
LumA
</td>
<td style="text-align:right;">
10.05
</td>
<td style="text-align:right;">
9.73
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;border-right:1px solid;">
MB-0005
</td>
<td style="text-align:center;">
Positive
</td>
<td style="text-align:center;">
Negative
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
LumB
</td>
<td style="text-align:right;">
10.04
</td>
<td style="text-align:right;">
9.73
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;border-right:1px solid;">
MB-0006
</td>
<td style="text-align:center;">
Positive
</td>
<td style="text-align:center;">
Negative
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
LumB
</td>
<td style="text-align:right;">
10.40
</td>
<td style="text-align:right;">
10.33
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;border-right:1px solid;">
MB-0008
</td>
<td style="text-align:center;">
Positive
</td>
<td style="text-align:center;">
Negative
</td>
<td style="text-align:center;">
ER+/HER2- High Prolif
</td>
<td style="text-align:center;">
LumB
</td>
<td style="text-align:right;">
11.28
</td>
<td style="text-align:right;">
9.96
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;border-right:1px solid;">
MB-0010
</td>
<td style="text-align:center;">
Positive
</td>
<td style="text-align:center;">
Negative
</td>
<td style="text-align:center;">
ER+/HER2- High Prolif
</td>
<td style="text-align:center;">
LumB
</td>
<td style="text-align:right;">
11.24
</td>
<td style="text-align:right;">
9.74
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;border-right:1px solid;">
MB-0014
</td>
<td style="text-align:center;">
Positive
</td>
<td style="text-align:center;">
Negative
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
LumB
</td>
<td style="text-align:right;">
10.79
</td>
<td style="text-align:right;">
9.28
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;border-right:1px solid;">
MB-0022
</td>
<td style="text-align:center;">
Positive
</td>
<td style="text-align:center;">
Negative
</td>
<td style="text-align:center;">
NA
</td>
<td style="text-align:center;">
claudin-low
</td>
<td style="text-align:right;">
10.44
</td>
<td style="text-align:right;">
8.61
</td>
</tr>
</tbody>
</table>
</div>
<div id="gene-expression-by-er-status" class="section level2 tabset">
<h2 class="tabset">Gene expression by ER status</h2>
<p>The figures below show gene expression distribution separated
according to ER status. The data are further split by HER2 status.</p>
<p>Four genes are shown:</p>
<ul>
<li>ESR1</li>
<li>ERBB2</li>
<li>GATA3</li>
<li>TP53</li>
</ul>
<pre><code>Code chunks: 1 chunk - set colours for plotting. 4 chunks - one for each plot.</code></pre>
<div id="esr1" class="section level3">
<h3>ESR1</h3>
<p><img src="week7_Rmarkdown_exercise_files/figure-html/boxESR1-1.png" width="576" /></p>
</div>
<div id="erbb2" class="section level3">
<h3>ERBB2</h3>
<p><img src="week7_Rmarkdown_exercise_files/figure-html/boxERBB2-1.png" width="576" /></p>
</div>
<div id="gata3" class="section level3">
<h3>GATA3</h3>
<p><img src="week7_Rmarkdown_exercise_files/figure-html/boxGATA3-1.png" width="576" /></p>
</div>
<div id="tp53" class="section level3">
<h3>TP53</h3>
<p><img src="week7_Rmarkdown_exercise_files/figure-html/boxTP53-1.png" width="576" /></p>
</div>
</div>
<div id="pam50-cancer-subtype" class="section level2">
<h2>PAM50 cancer subtype</h2>
<p>The PAM50 classification of breast cancer included 5 basic subtypes.
An additional “claudin low” subtype is also included in metabric and
some tumours are not classified (NC). The seven categories included in
metabric are:</p>
<table>
<thead>
<tr class="header">
<th align="left">Category</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Basal</td>
<td align="left">Basal-like</td>
</tr>
<tr class="even">
<td align="left">Her2</td>
<td align="left">HER2 enriched</td>
</tr>
<tr class="odd">
<td align="left">LumA</td>
<td align="left">Luminal A</td>
</tr>
<tr class="even">
<td align="left">LumB</td>
<td align="left">Luminal B</td>
</tr>
<tr class="odd">
<td align="left">NC</td>
<td align="left">Not classified</td>
</tr>
<tr class="even">
<td align="left">Normal</td>
<td align="left">Normal-like</td>
</tr>
<tr class="odd">
<td align="left">claudin-low</td>
<td align="left">Claudin Low</td>
</tr>
</tbody>
</table>
<p>The table below shows the mean gene expression for 8 genes across all
samples according to their PAM50 cancer subtype.</p>
<pre><code>Code chunk: summarise expression data and generate table</code></pre>
<table class="table table-striped table-hover" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
PAM50
</th>
<th style="text-align:right;">
ESR1
</th>
<th style="text-align:right;">
ERBB2
</th>
<th style="text-align:right;">
PGR
</th>
<th style="text-align:right;">
TP53
</th>
<th style="text-align:right;">
PIK3CA
</th>
<th style="text-align:right;">
GATA3
</th>
<th style="text-align:right;">
FOXA1
</th>
<th style="text-align:right;">
MLPH
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;font-weight: bold;border-right:1px solid;">
claudin-low
</td>
<td style="text-align:right;">
6.68
</td>
<td style="text-align:right;">
9.69
</td>
<td style="text-align:right;">
5.45
</td>
<td style="text-align:right;">
6.23
</td>
<td style="text-align:right;">
5.95
</td>
<td style="text-align:right;">
7.40
</td>
<td style="text-align:right;">
9.47
</td>
<td style="text-align:right;">
9.48
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;border-right:1px solid;">
LumA
</td>
<td style="text-align:right;">
10.93
</td>
<td style="text-align:right;">
10.65
</td>
<td style="text-align:right;">
6.67
</td>
<td style="text-align:right;">
6.20
</td>
<td style="text-align:right;">
5.90
</td>
<td style="text-align:right;">
10.43
</td>
<td style="text-align:right;">
11.65
</td>
<td style="text-align:right;">
12.25
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;border-right:1px solid;">
LumB
</td>
<td style="text-align:right;">
11.16
</td>
<td style="text-align:right;">
10.45
</td>
<td style="text-align:right;">
6.14
</td>
<td style="text-align:right;">
6.18
</td>
<td style="text-align:right;">
5.97
</td>
<td style="text-align:right;">
10.29
</td>
<td style="text-align:right;">
11.61
</td>
<td style="text-align:right;">
11.98
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;border-right:1px solid;">
Her2
</td>
<td style="text-align:right;">
7.02
</td>
<td style="text-align:right;">
13.07
</td>
<td style="text-align:right;">
5.40
</td>
<td style="text-align:right;">
6.03
</td>
<td style="text-align:right;">
5.96
</td>
<td style="text-align:right;">
8.78
</td>
<td style="text-align:right;">
11.14
</td>
<td style="text-align:right;">
11.55
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;border-right:1px solid;">
Normal
</td>
<td style="text-align:right;">
9.76
</td>
<td style="text-align:right;">
10.54
</td>
<td style="text-align:right;">
5.97
</td>
<td style="text-align:right;">
6.14
</td>
<td style="text-align:right;">
5.85
</td>
<td style="text-align:right;">
9.67
</td>
<td style="text-align:right;">
11.12
</td>
<td style="text-align:right;">
11.90
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;border-right:1px solid;">
Basal
</td>
<td style="text-align:right;">
6.00
</td>
<td style="text-align:right;">
9.91
</td>
<td style="text-align:right;">
5.36
</td>
<td style="text-align:right;">
6.24
</td>
<td style="text-align:right;">
6.08
</td>
<td style="text-align:right;">
7.15
</td>
<td style="text-align:right;">
6.72
</td>
<td style="text-align:right;">
8.21
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;border-right:1px solid;">
NC
</td>
<td style="text-align:right;">
10.83
</td>
<td style="text-align:right;">
10.23
</td>
<td style="text-align:right;">
6.25
</td>
<td style="text-align:right;">
6.22
</td>
<td style="text-align:right;">
5.66
</td>
<td style="text-align:right;">
10.29
</td>
<td style="text-align:right;">
11.18
</td>
<td style="text-align:right;">
11.92
</td>
</tr>
</tbody>
</table>
</div>
<div id="three-gene-classifier" class="section level2">
<h2>Three gene classifier</h2>
<p>Here we look at GATA3 and ESR1 expression according to the three gene
classifier.</p>
<pre><code>Code chunk: 4 plots arranged using `patchwork`. Caption in chunk options.</code></pre>
<div class="figure">
<img src="week7_Rmarkdown_exercise_files/figure-html/threeGC-1.png" alt="Expression of ESR1 and GATA3 according to 3-gene classifier. (a) Bar plot showing numbers of samples in each category. (b) Scatter plot showing ESR1 expression against GAT3 expression for each sample. (c) Violin plot showing distribution of ESR1 expression. (d) Violin plot showing distribution of GATA3 expression." width="1344" />
<p class="caption">
Expression of ESR1 and GATA3 according to 3-gene classifier. (a) Bar
plot showing numbers of samples in each category. (b) Scatter plot
showing ESR1 expression against GAT3 expression for each sample. (c)
Violin plot showing distribution of ESR1 expression. (d) Violin plot
showing distribution of GATA3 expression.
</p>
</div>
<hr />
<p>END OF REPORT</p>
</div>
<div id="session-info" class="section level2">
<h2>Session Info</h2>
<pre class="r"><code>devtools::session_info()</code></pre>
<pre><code>## ─ Session info ───────────────────────────────────────────────────────────────
##  setting  value
##  version  R version 4.5.0 (2025-04-11)
##  os       Ubuntu 24.04.2 LTS
##  system   x86_64, linux-gnu
##  ui       X11
##  language (EN)
##  collate  en_US.UTF-8
##  ctype    en_US.UTF-8
##  tz       Europe/London
##  date     2025-04-25
##  pandoc   3.1.3 @ /usr/bin/ (via rmarkdown)
##  quarto   NA
## 
## ─ Packages ───────────────────────────────────────────────────────────────────
##  package     * version date (UTC) lib source
##  bit           4.6.0   2025-03-06 [1] CRAN (R 4.5.0)
##  bit64         4.6.0-1 2025-01-16 [1] CRAN (R 4.5.0)
##  bslib         0.9.0   2025-01-30 [1] CRAN (R 4.5.0)
##  cachem        1.1.0   2024-05-16 [1] CRAN (R 4.5.0)
##  cli           3.6.4   2025-02-13 [1] CRAN (R 4.5.0)
##  colorspace    2.1-1   2024-07-26 [1] CRAN (R 4.5.0)
##  crayon        1.5.3   2024-06-20 [1] CRAN (R 4.5.0)
##  devtools      2.4.5   2022-10-11 [1] CRAN (R 4.5.0)
##  digest        0.6.37  2024-08-19 [1] CRAN (R 4.5.0)
##  dplyr       * 1.1.4   2023-11-17 [1] CRAN (R 4.5.0)
##  ellipsis      0.3.2   2021-04-29 [1] CRAN (R 4.5.0)
##  evaluate      1.0.3   2025-01-10 [1] CRAN (R 4.5.0)
##  farver        2.1.2   2024-05-13 [1] CRAN (R 4.5.0)
##  fastmap       1.2.0   2024-05-15 [1] CRAN (R 4.5.0)
##  forcats     * 1.0.0   2023-01-29 [1] CRAN (R 4.5.0)
##  fs            1.6.6   2025-04-12 [1] CRAN (R 4.5.0)
##  generics      0.1.3   2022-07-05 [1] CRAN (R 4.5.0)
##  ggplot2     * 3.5.2   2025-04-09 [1] CRAN (R 4.5.0)
##  glue          1.8.0   2024-09-30 [1] CRAN (R 4.5.0)
##  gtable        0.3.6   2024-10-25 [1] CRAN (R 4.5.0)
##  hms           1.1.3   2023-03-21 [1] CRAN (R 4.5.0)
##  htmltools     0.5.8.1 2024-04-04 [1] CRAN (R 4.5.0)
##  htmlwidgets   1.6.4   2023-12-06 [1] CRAN (R 4.5.0)
##  httpuv        1.6.16  2025-04-16 [1] CRAN (R 4.5.0)
##  jquerylib     0.1.4   2021-04-26 [1] CRAN (R 4.5.0)
##  jsonlite      2.0.0   2025-03-27 [1] CRAN (R 4.5.0)
##  kableExtra  * 1.4.0   2024-01-24 [1] CRAN (R 4.5.0)
##  knitr       * 1.50    2025-03-16 [1] CRAN (R 4.5.0)
##  labeling      0.4.3   2023-08-29 [1] CRAN (R 4.5.0)
##  later         1.4.2   2025-04-08 [1] CRAN (R 4.5.0)
##  lifecycle     1.0.4   2023-11-07 [1] CRAN (R 4.5.0)
##  lubridate   * 1.9.4   2024-12-08 [1] CRAN (R 4.5.0)
##  magrittr      2.0.3   2022-03-30 [1] CRAN (R 4.5.0)
##  memoise       2.0.1   2021-11-26 [1] CRAN (R 4.5.0)
##  mime          0.13    2025-03-17 [1] CRAN (R 4.5.0)
##  miniUI        0.1.2   2025-04-17 [1] CRAN (R 4.5.0)
##  munsell       0.5.1   2024-04-01 [1] CRAN (R 4.5.0)
##  patchwork   * 1.3.0   2024-09-16 [1] CRAN (R 4.5.0)
##  pillar        1.10.2  2025-04-05 [1] CRAN (R 4.5.0)
##  pkgbuild      1.4.7   2025-03-24 [1] CRAN (R 4.5.0)
##  pkgconfig     2.0.3   2019-09-22 [1] CRAN (R 4.5.0)
##  pkgload       1.4.0   2024-06-28 [1] CRAN (R 4.5.0)
##  profvis       0.4.0   2024-09-20 [1] CRAN (R 4.5.0)
##  promises      1.3.2   2024-11-28 [1] CRAN (R 4.5.0)
##  purrr       * 1.0.4   2025-02-05 [1] CRAN (R 4.5.0)
##  R6            2.6.1   2025-02-15 [1] CRAN (R 4.5.0)
##  Rcpp          1.0.14  2025-01-12 [1] CRAN (R 4.5.0)
##  readr       * 2.1.5   2024-01-10 [1] CRAN (R 4.5.0)
##  remotes       2.5.0   2024-03-17 [1] CRAN (R 4.5.0)
##  rlang         1.1.6   2025-04-11 [1] CRAN (R 4.5.0)
##  rmarkdown     2.29    2024-11-04 [1] CRAN (R 4.5.0)
##  rstudioapi    0.17.1  2024-10-22 [1] CRAN (R 4.5.0)
##  sass          0.4.10  2025-04-11 [1] CRAN (R 4.5.0)
##  scales        1.3.0   2023-11-28 [1] CRAN (R 4.5.0)
##  sessioninfo   1.2.3   2025-02-05 [1] CRAN (R 4.5.0)
##  shiny         1.10.0  2024-12-14 [1] CRAN (R 4.5.0)
##  stringi       1.8.7   2025-03-27 [1] CRAN (R 4.5.0)
##  stringr     * 1.5.1   2023-11-14 [1] CRAN (R 4.5.0)
##  svglite       2.1.3   2023-12-08 [1] CRAN (R 4.5.0)
##  systemfonts   1.2.2   2025-04-04 [1] CRAN (R 4.5.0)
##  tibble      * 3.2.1   2023-03-20 [1] CRAN (R 4.5.0)
##  tidyr       * 1.3.1   2024-01-24 [1] CRAN (R 4.5.0)
##  tidyselect    1.2.1   2024-03-11 [1] CRAN (R 4.5.0)
##  tidyverse   * 2.0.0   2023-02-22 [1] CRAN (R 4.5.0)
##  timechange    0.3.0   2024-01-18 [1] CRAN (R 4.5.0)
##  tzdb          0.5.0   2025-03-15 [1] CRAN (R 4.5.0)
##  urlchecker    1.0.1   2021-11-30 [1] CRAN (R 4.5.0)
##  usethis       3.1.0   2024-11-26 [1] CRAN (R 4.5.0)
##  vctrs         0.6.5   2023-12-01 [1] CRAN (R 4.5.0)
##  viridisLite   0.4.2   2023-05-02 [1] CRAN (R 4.5.0)
##  vroom         1.6.5   2023-12-05 [1] CRAN (R 4.5.0)
##  withr         3.0.2   2024-10-28 [1] CRAN (R 4.5.0)
##  xfun          0.52    2025-04-02 [1] CRAN (R 4.5.0)
##  xml2          1.3.8   2025-03-14 [1] CRAN (R 4.5.0)
##  xtable        1.8-4   2019-04-21 [1] CRAN (R 4.5.0)
##  yaml          2.3.10  2024-07-26 [1] CRAN (R 4.5.0)
## 
##  [1] /home/sawle01/R/x86_64-pc-linux-gnu-library/4.5
##  [2] /usr/local/lib/R/site-library
##  [3] /usr/lib/R/site-library
##  [4] /usr/lib/R/library
##  * ── Packages attached to the search path.
## 
## ──────────────────────────────────────────────────────────────────────────────</code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
